File: solutions/cpp/Problem7.cpp
#pragma once
/*
7. Reverse Integer

Given a 32-bit signed integer, reverse digits of an integer.
*/
#include <iostream>
#include <vector>
#include <string>


class Solution {
public:
	int reverse(int x)	{
		long long result = 0;
		while (x != 0)
		{
			result *= 10;
			result += x % 10;
			x /= 10;
		}
		if (result > INT_MAX || result < INT_MIN)
		{
			return 0;
		}
		return (int) result;
	}
	void test(){
		std::cout << "Reversing 387: " << reverse(387) << std::endl;
		std::cout << "Reversing -145: " << reverse(-145) << std::endl;
		std::cout << "Reversing 2147483648: " << reverse(2147483648) << std::endl;
	}
}; 

File: solutions/python3/problem7.py
# -*- coding: utf-8 -*-
"""
7. Reverse Integer

Given a 32-bit signed integer, reverse digits of an integer.

Note:
Assume we are dealing with an environment which could only store integers within the 32-bit signed integer range: [−231,  231 − 1].
For the purpose of this problem, assume that your function returns 0 when the reversed integer overflows.
"""


class Solution:
    def reverse(self, x: int) -> int:
        if x < 0:
            b_neg = True
            x = -x
            num_thres = 2 ** 31
        else:
            b_neg = False
            num_thres = 2 ** 31 - 1

        res = 0
        while x > 0:
            tmp = x % 10
            if res > ((num_thres - tmp) / 10):
                return 0

            res = res * 10 + tmp
            x = x // 10

        if b_neg:
            return -res
        else:
            return res
